docs/rfcs/20220131_raft_snapshots.md

```diff
@@ -0,0 +1,150 @@
+
+The main idea is to use existing log replication to ship the snapshot. It's easier to implement, but just like the reconfiguration batches, it shifts the offsets and requires that we continue to rely on offset translation.
+
+### Overview
+
+When a leader decides to do compaction at offset X, it:
+ 
+  1. saves the current snapshot
+  2. waits until it's safe to do compaction:
+    a. segment containing X is closed
+    b. tx front is passed the end of the segment
+  3. replicates the snapshot
+ 
+When a leader or a follower sees the snapshot via `state_machine::apply` (this only happens when the batch is replicated to the majority and can't be lost), it:
+
+  - performs the compaction when the current node is the leader and it sent the snapshot (within the current term)
+  - when the snapshot isn't applied:
+    1. rewinds log at offset X
+    2. applies the snapshot
+  - when the snapshot is applied:
+    - performs the compaction
+
+There is a problem with this approach: a follower doesn't know if the batches it replays contain holes, so its current state may be incorrect until it sees the snapshot batch. This becomes a problem when Redpanda supports reads from replicas. There are several ways to handle this: 
+
+#### Block replica reads
+
+A follower is allowed to serve reads only up to the last observed snapshot. Even when the prior batches contain holes, they become visible when the snapshot is applied and corrected.
+
+#### Modify the `append_entries_request`
+
+We can modify the `append_enties_request` to include an offset of the last compaction. When a follower sees that last compaction is less than the current offset, it knows that the latest batches don't have holes and it's safe to serve the requests.
+
+When we modify `append_entries_request` or when we introduce and switch to the new API, we need to take upgrades into account. In order to avoid shutting down a cluster for an update, we need to put the change behind a feature flag, update the brokers one by one, and only then active the feature.
+
+## Approach #2: Ship the snapshot (cached state) via the new API
+
+Alternatively, we could use a side channel instead of the main replication path to ship the snapshot. With this approach we don't shift the offset and it eventually opens a way to get rid of the offset translation layer.
+
+### Overview
+
+When a leader decides to compact the log at offset X, it:
+
+  1. makes a snapshot and sends it to a majority of the nodes via the new API
+  2. waits until it's safe to do compaction:
+    a. the segment containing X is closed
+    b. tx front is passed the end of the segment
+  3. starts include the new offset X to the `append_entries_request`
+  4. performs the compaction
+ 
+When a follower observes a compacted offset past its snapshot via `append_entries_request`, it:
+
+  1. queries all nodes to find a node that has the snapshot (this is a majority of the nodes, so the availability is the same as with Raft)
```

Comment:
An idea: can we repurpose `InstallSnapshot` rpc to do this? Something like:
* the follower detects (based on arrived `append_entries_request`) if log entries needed to advance particular state machines have already been deleted/compacted on the leader. If this is the case, it includes a request for corresponding snapshots in the `append_entries_reply`.
* the leader then sends an InstallSnapshot request with needed snapshots.